pup-0003.md

```diff
@@ -0,0 +1,154 @@
+are merged to release branches which then are merged to newer release branches
+and then to master (a process called "merging forward"). More info about
+merging forward can be found in our documentation:
+
+http://docs.pulpproject.org/en/2.12/dev-guide/contributing/merging.html#merging-your-changes
+
+Also, a good illustration of this process is outlined in the following webpage:
+
+http://nvie.com/posts/a-successful-git-branching-model/
+
+This PUP is to use cherry-picking instead of merging forward. Using this model,
+PRs would be made against master and then changes would be cherry-picked to
+release branches as needed. For a more detailed workflow, see the Design
+section.
+
+
+## Motivation
+
+There are a few motivating factors for this proposal. First, our git history is
+complex with multiple branches feeding into one another and numerous merge
+commits. Using cherry-picks would simplify our git history by keeping it more
+linear and eliminating merge commits.
+
+Secondly, opening a PR requires knowledge of our current process. This can
+create a barrier to entry for community members. With the cherry-pick model,
+all PRs are opened against master and the onus to update release branches is on
+the person merging (who usually has better knowledge of which branches a commit
+needs to be applied to).
+
+Another minor caveat is that changes that we want to only be made to older
+release branches (but not to newer ones) still need to be merged forward. We
+could elminate the need for this with cherry-picking. It's worth noting that
+this rarely happens.
+
+Lastly, we've also seen some mistakes where branches are accidentally merged
+into other branches. Commits could be accidentally cherry-picked but this would
+be less disastrous since it's only a single commit.
+
+
+## Detailed Design
+
+This proposal would keep the current branching scheme [1] involving master,
+x.y-dev branches, and x.y-release branhces. The workflow by which changes are
+moved between the branches would however change.
+
+### Story
+
+1. All story PRs are opened and merged against master.
+
+### Bug Fix
+
+1. All bug fix PRs are also opened against master
+2. After the PR is ready to be merged, the person merging the commit would
+   first merge the commit to master. Then, this person would cherry-pick the
+   change from master to the latest x.y-dev branch.
+3. The person cherry-picking would also need to track in redmine which release
+   branches the change has been made to.
+
+### Hotfix
+
+The current hotfix process [2] would remain basically unchanged. However,
+instead of merging forward the commit, we'd use cherry-picking.
+
+1. The affected release would be branched and a hotfix would be applied.
+2. After the release, the branch gets merged into the x.y-release branch.
+3. The change is then cherry-picked into the x.y-dev branch and then
+   cherry-picked forward to any other x.y-dev branches and master.
```

Comment:
My understanding of the current process is that we would only release against the current release stream ever. It is important for users to stay up to date. As Pulp is packaged in various distros those packagers could choose to apply the security fix onto their older, packaged version of Pulp.